### PR TITLE
Add utf8proc port

### PR DESCRIPTION
**Pull request by trbabb**
This adds [utf8proc](https://github.com/JuliaStrings/utf8proc), a lightweight MIT-licensed C library for processing Unicode.

Please note that the hash used for the zipped repository file was taken from the "unexpected hash" error message emitted by em++ when I attempted to build a test project using `-s USE_UTF8PROC`. When following the normal directions of piping `curl` to `shasum -a 512`, I get a different result. I'm guessing this is an issue with the hash method used by emscripten (e.g., it's not hashing the bytes in a way that exactly reproduces `shasum`), but let me know if I should be doing something different. (Can others reproduce the included hash in the terminal?)

### CODE REVIEW

**Comment by kripken** [comment]
Hi @trbabb , and thanks for the PR!

In general we've only been adding high-priority ports recently because we have some general ideas of changing the ports system, either to use some external tool, or to make it more pluggable. But with that said, I'm not sure if anyone is working on those ideas atm.

I haven't heard of utf8proc before - is it a widely-used library?

**Comment by trbabb** [comment]
It has >800 stars on GitHub, and it's the basis for unicode support in the Julia language, and the doc page alludes to its use in Ruby and PostgreSQL. It's [actively maintained](https://github.com/JuliaStrings/utf8proc/issues/246).

**Comment by trbabb** [comment]
Is there a more preferred way to handle unicode text in a standards compliant way in an emscripten-compiled app?

**Comment by sbc100** [comment]
> Is there a more preferred way to handle unicode text in a standards compliant way in an emscripten-compiled app?

We have support for libicu (-sUSE_ICU).. does that do what you want?

Also, even if we don't have all the libraries you need, the general idea is that you can take any library you like and compile it yourself along with your project.  I'm not sure what the bar should be for adding new ports, but I'd like to keep it pretty high.  

Is there some reason you think this library makes sense as a builtin port like this?  Do you not use other libraries too that are not ports?

**Comment by trbabb** [comment]
I'm following the [recommendation in the documentation](https://emscripten.org/docs/compiling/Building-Projects.html#adding-more-ports) which suggests that adding more ports is welcome.

From a developer perspective, adding `-s USE_SOME_PORT` is a zillion times more ergonomic than wrangling a separate library from scratch and trying to get it to compile on a novel architecture and toolchain. In my own case, it also makes my build system and repo much simpler when I can handle the library by adding a single flag.

Adding a port also means that developers aren't (often erroneously) recreating each others' work over and over in private. It took me a couple hours to figure out what `utf8proc` actually needed; I'd be happy to save other devs that time. And also share updates and bug / configuration fixes.

I guess I don't see the advantage of being restrictive. A rich, easy-to-use ecosystem is a big advantage for a toolchain, and it seems like there's a great start here!

Either way, it's up to you guys!

**Comment by trbabb** [comment]
Also, w.r.t `icu`— I didn't recognize the lib by name in the list. FWIW, the org I'm in (not related to this specific addition... for now) decided to forgo including `icu` because it makes the bundle substantially bigger. This library is a lot lighter.

**Comment by sbc100** [comment]
> I'm following the [recommendation in the documentation](https://emscripten.org/docs/compiling/Building-Projects.html#adding-more-ports) which suggests that adding more ports is welcome.

Oh, I think I should have updated that documentation.

> 
> From a developer perspective, adding `-s USE_SOME_PORT` is a zillion times more ergonomic than wrangling a separate library from scratch and trying to get it to compile on a novel architecture and toolchain. In my own case, it also makes my build system and repo much simpler when I can handle the library by adding a single flag.
> 
> Adding a port also means that developers aren't (often erroneously) recreating each others' work over and over in private. It took me a couple hours to figure out what `utf8proc` actually needed; I'd be happy to save other devs that time. And also share updates and bug / configuration fixes.
> 
> I guess I don't see the advantage of being restrictive. A rich, easy-to-use ecosystem is a big advantage for a toolchain, and it seems like there's a great start here!

Yes, I completely agree that a rich ports system is awesome and I really want to have that for emscripten, but its also something perhaps makes sense outside of the core repo, or at least doesn't have to block progress on emscripten itself.   Imagine if we had 1000+ ports (very common for ports system) and they all lived upstream here in the emscripten repo, we would have to have some way of making progress without having to ensure that all the ports build with every change to emscripten itself its not really feasible to do all the testing during CI.  Normally the way that works is that each port has its own maintainer, and they update it when they can. 

Finally, I'm not sure we want to evolve our existing ports system to something that supports 1000+ packages when there are other tools out there that are already built for that.  For example https://vcpkg.io/en/index.html is one that I find very compelling.

In other words, we do want to have a rick ports system, but I'm not sure that simply incrementally growing our in house one is that most sensible way to go.


> 
> Either way, it's up to you guys!

**Comment by sbc100** [comment]
We now have a contrib ports directory if you want to add this there.

Closing for now, feel free to re-open.
